ocaml-lsp-server/src/code_actions/action_mark_remove_unused.ml

```diff
@@ -127,27 +172,226 @@ let code_action_remove_value doc pos (diagnostic : Diagnostic.t) =
       |> Option.map ~f:(fun range ->
              code_action_remove_range doc diagnostic range))
 
-let find_unused_diagnostic pos (ds : Diagnostic.t list) =
-  List.find ds ~f:(fun d ->
-      let in_range =
-        match Position.compare_inclusion pos d.range with
-        | `Outside _ -> false
-        | `Inside -> true
-      in
-      in_range && Diagnostic_util.is_unused_var_warning d.message)
+(** [create_mark_action ~title doc pos d] creates a code action that resolves
+    the diagnostic [d] by inserting an underscore at [pos] in [doc]. *)
+let create_mark_action ~title doc pos d =
+  let edit =
+    Document.edit
+      doc
+      [ { range = Range.create ~start:pos ~end_:pos; newText = "_" } ]
+  in
+  CodeAction.create
+    ~diagnostics:[ d ]
+    ~title
+    ~kind:CodeActionKind.QuickFix
+    ~edit
+    ~isPreferred:true
+    ()
+
+let action_mark_type doc pos (d : Diagnostic.t) =
+  let open Option.O in
+  Document.Merlin.with_pipeline_exn (Document.merlin_exn doc) (fun pipeline ->
+      enclosing_pos pipeline pos
+      |> List.find_map ~f:(fun (_, node) ->
+             match node with
+             | Browse_raw.Type_declaration
+                 { typ_name = { loc = { loc_start; _ }; _ }; _ } ->
+               Some loc_start
+             | _ -> None))
+  |> Fiber.map ~f:(fun m_name_loc_start ->
+         let* name_loc_start = m_name_loc_start in
+         let+ start = Position.of_lexical_position name_loc_start in
+         create_mark_action ~title:"Mark type as unused" doc start d)
+
+let contains loc pos =
+  match Position.compare_inclusion pos (Range.of_loc loc) with
+  | `Outside _ -> false
+  | `Inside -> true
+
+let action_mark_for_loop_index doc pos (d : Diagnostic.t) =
+  let open Option.O in
+  let module I = Ocaml_typing.Tast_iterator in
+  let exception Found of Warnings.loc in
+  let iterator =
+    let expr iter (e : Typedtree.expression) =
```

Comment:
I might be wrong here, but can't you iterate over the parsetree instead?

Reply:
Yes, and I'm not sure why I didn't. I'll change that.